### PR TITLE
Add jflip module

### PR DESCRIPTION
**Pull request by cindywxw**

### CODE REVIEW

**Comment by gfetterman** [comment]
There are three binary files in this PR. Can you modify `.gitignore` to remove them?

**Comment by gfetterman** [comment]
There's a change to the top-level `SConstruct` file that seems to consist of the (unnecessary) addition of a single space. Including the change clutters the git history and the PR. Can you change it back?

**Comment by gfetterman** [comment]
There's also a single-line change to `jill_module_skel.cc`. It isn't relevant to jflip, so please revert it.

File: modules/jflip.cc [hunk 1]
```diff
@@ -0,0 +1,291 @@
+/*
```
Comment (gfetterman) [inline modules/jflip.cc]: This comment is no longer accurate - `jflip` isn't a skeleton any more. Please remove the first paragraph and the list and replace with a description of the module's function.

It's really important that documentation, including comments in source code, match the behavior of the program.

File: modules/jflip.cc [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Copyright (C) 2010-2013 C Daniel Meliza <dan || meliza.org>
+ */
+#include <iostream>
+#include <signal.h>
+#include <boost/shared_ptr.hpp>
+
+// #include <sys/time.h>
```
Comment (gfetterman) [inline modules/jflip.cc]: This can be removed.

File: modules/jflip.cc [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+#include <stdlib.h>     /* srand, rand */
+#include <time.h>       /* time */
+#include <stdexcept>
+
+#include "jill/logging.hh"
+#include "jill/jack_client.hh"
+#include "jill/program_options.hh"
+
+// https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/tree/master/2013/08/14/cpp
```
Comment (gfetterman) [inline modules/jflip.cc]: You aren't using the more complex randomization algorithm any more, so these includes should be removed, not commented out.

File: modules/jflip.cc [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    string server_name;
+    /** The client name (used in internal JACK representations) */
+    string client_name;
+
+
+    /** A vector of inputs to connect to the client */
+    std::vector<string> input_ports;
+    /** A vector of outputs to connect to the client */
+    std::vector<string> output_ports;
+        /** The MIDI output channel */
```
Comment (gfetterman) [inline modules/jflip.cc]: `jflip` doesn't have a MIDI output port, so these lines can be removed.

File: modules/jflip.cc [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+protected:
+
+    virtual void print_usage();
+
+}; // jflip_options
+
+static jflip_options options(PROGRAM_NAME);
+static boost::shared_ptr<jack_client> client;
+// boost::shared_ptr<dsp::crossing_trigger<sample_t> > trigger;
```
Comment (gfetterman) [inline modules/jflip.cc]: These commented lines aren't needed, so should be removed.

File: modules/jflip.cc [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+// boost::shared_ptr<dsp::buffered_data_writer> arf_thread;
+jack_port_t *port_in, *port_out, *port_trig,*port_count;
+static int ret = EXIT_SUCCESS;
+static int running = 1;
+static int stopping = 0;
+static int xruns = 0;
+
+/* data storage for event times */
```
Comment (gfetterman) [inline modules/jflip.cc]: If this isn't needed, it should be removed.

File: modules/jflip.cc [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+int
+process(jack_client *client, nframes_t nframes, nframes_t)
+{
+    sample_t *in = client->samples(port_in, nframes);
+    sample_t *out = client->samples(port_out, nframes);
+
+    nframes_t inputSize = nframes;
+
+    memcpy(out, in, nframes * sizeof(sample_t));
+    /* rand() method */
```
Comment (gfetterman) [inline modules/jflip.cc]: This comment is probably unnecessary - it's clear enough from the loop code what's happening.

File: modules/jflip.cc [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    int i = 0;
+    std::srand(std::time(NULL));
+    while( i < inputSize) { 
+        int coin = rand()%2;
+        if (coin == 0) *(out+i) *= -1;
```
Comment (gfetterman) [inline modules/jflip.cc]: This style can be easy to misread for anything more complex than `if (...) break` or `if (...) continue`. Can you bring the multiplication step down onto its own line?

File: modules/jflip.cc [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        i++;
+    }
+
+    /* use Lemire's picking number method */
```
Comment (gfetterman) [inline modules/jflip.cc]: If you aren't using this method any more, it's better to remove it than comment it out.

File: modules/jflip.cc [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        return 0;
+}
+
+
+/** handle xrun events */
+int
+jack_xrun(jack_client *client, float delay)
+{
+        __sync_add_and_fetch(&xruns, 1); // gcc specific
+        // arf_thread->xrun();
```
Comment (gfetterman) [inline modules/jflip.cc]: `jflip` isn't writing to an ARF file, so this can be removed.

File: modules/jflip.cc [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+/** handle server shutdowns */
+void
+jack_shutdown(jack_status_t code, char const *)
+{
+        ret = -1;
+        LOG << "jackd shut the client down";
+        __sync_add_and_fetch(&stopping, 1);
+        // wait for at least one process loop; not strictly async safe
+        usleep(2e6 * client->buffer_size() / client->sampling_rate());
+        // if (arf_thread) {
```
Comment (gfetterman) [inline modules/jflip.cc]: `jflip` isn't writing to an ARF file, so this can be removed.

File: modules/jflip.cc [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+        running = 0;
+}
+
+
+/** handle POSIX signals */
+void
+signal_handler(int sig)
+{
+        ret = sig;
+        // if (arf_thread) {
```
Comment (gfetterman) [inline modules/jflip.cc]: `jflip` isn't writing to an ARF file, so this can be removed.

File: modules/jflip.cc [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+
+                // start client
+                client.reset(new jack_client(options.client_name, options.server_name));
+
+                // register ports
+                port_in = client->register_port("in",JACK_DEFAULT_AUDIO_TYPE,
+                                                JackPortIsInput, 0);
+                port_out = client->register_port("out", JACK_DEFAULT_AUDIO_TYPE,
+                                                 JackPortIsOutput, 0);
+                // port_trig = client->register_port("trig_out",JACK_DEFAULT_MIDI_TYPE,
```
Comment (gfetterman) [inline modules/jflip.cc]: If this isn't needed, remove it.

File: modules/jflip.cc [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        signal(SIGHUP,  signal_handler);
+
+                // register jack callbacks
+                client->set_shutdown_callback(jack_shutdown);
+                client->set_xrun_callback(jack_xrun);
+                client->set_process_callback(process);
+
+                client->set_buffer_size_callback(jack_bufsize);
+
+                // uncomment if you need these callbacks
```
Comment (gfetterman) [inline modules/jflip.cc]: If these aren't needed, remove them.

File: modules/jflip.cc [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+                // client->set_buffer_size_callback(jack_bufsize);
+                // jack_set_latency_callback (client->client(), jack_latency, 0);
+
+                // activate client
+                client->activate();
+                // arf_thread->start();
```
Comment (gfetterman) [inline modules/jflip.cc]: This isn't needed; remove it.

File: modules/jflip.cc [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+        // this section is for general JILL options. try to maintain consistency
+        // with other modules
+        po::options_description jillopts("JILL options");
+        jillopts.add_options()
+                ("server,s",  po::value<string>(&server_name), "connect to specific jack server")
+                ("name,n",    po::value<string>(&client_name)->default_value(_program_name),
+                 "set client name")
+                ("in,i",      po::value<stringvec>(), "add connection to input port")
+                ("out,o",     po::value<stringvec>(), "add connection to output port")
+                // ("buffer",     po::value<float>(&buffer_size_s)->default_value(2.0), "minimum ringbuffer size (s)")
```
Comment (gfetterman) [inline modules/jflip.cc]: If this isn't needed, remove it.

**Comment by cindywxw** [comment]
Thanks for your comments, Graham. I changed the jflip.cc and changed the lines in SConstruct and jill_module_skel.cc back. I'm not very familiar git commands so I'm not sure if I'm doing it right. Maybe you can help me with those files when we meet?

File: modules/jflip.cc [hunk 17]
```diff
@@ -0,0 +1,228 @@
+/*
+ * jflip - whiten the input data by randomly inverting it around
+ *
```
Comment (gfetterman) [inline modules/jflip.cc]: Can you add the license text back in (you can copy it from another module)?

Comment (gfetterman) [reply modules/jflip.cc]: Also, the program description looks cut off. Randomly inverting it around what?

**Comment by gfetterman** [comment]
Looks good. Thanks for making those changes.

**Pull request merged**